src/mpi/attr/attr_impl.c

```diff
@@ -98,9 +98,11 @@ int MPIR_Type_create_keyval_impl(MPI_Type_copy_attr_function * type_copy_attr_fn
     keyval_ptr->was_freed = 0;
     keyval_ptr->kind = MPIR_DATATYPE;
     keyval_ptr->extra_state = extra_state;
-    keyval_ptr->copyfn.user_function = type_copy_attr_fn;
+    /* add (void *) cast since we are assigning MPI_Type_copy_attr_function to
+     * MPI_Comm_copy_attr_function */
+    keyval_ptr->copyfn.user_function = (void *) type_copy_attr_fn;
```

Comment:
I'm surprised our warning tests don't flag this cast. Is there any way to avoid it?

Reply:
That's just standard C, `void*` may be converted to and from any other pointer type. The only option I know that would warn under this usage is `-Wc++-compat`, which I personally use in my own C codes.

Reply:
BTW, the proper cast in this case should be
```
keyval_ptr->copyfn.user_function = (MPI_Comm_copy_attr_function) (void(*)(void)) type_copy_attr_fn
```
The `(void(*)(void))` is required to prevent warnings about cast with incompatible types, I do not remember what's the option that triggers such warnings (or maybe the warning occurs when using clang, not GCC).

Reply:
> That's just standard C, `void*` may be converted to and from any other pointer type. The only option I know that would warn under this usage is `-Wc++-compat`, which I personally use in my own C codes.

GCC will warn that ISO C forbids it:
```
warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
```

Reply:
Oh, yes, sorry, I messed up the "funciton" vs "object" distinction. Still, I believe the cast line I provided above is correct.

Reply:
Code for reference:
```c
#include <stdio.h>
int f(int n)
{
    printf("%d\n", n);
    return n*n;
}
int main(void)
{
    int (*p)(int) = (void *)f;
    int x = p(7);

    return 0;
}
```
```
raffenet@csi0363877:/tmp/ $ gcc-13 -Wpedantic foo.c
foo.c: In function 'main':
foo.c:9:21: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
    9 |     int (*p)(int) = (void *)f;
      |                     ^
foo.c:9:21: warning: ISO C forbids initialization between function pointer and 'void *' [-Wpedantic]
```

Reply:
> Oh, yes, sorry, I messed up the "funciton" vs "object" distinction. Still, I believe the cast line I provided above is correct.

If it is still an issue with the new cast, we could just add a FIXME comment for now. This is unlikely to cause any issue for the ABI target platform.

Reply:
I'll rewrite it into a  void function cast.

Reply:
[I have to complain] We are casting to tell the compiler to not to check and not to warn. Both `(void *)` and `(void (*)(void))` or `(MPI_Comm_copy_attr_function)` says no more and no less. So why don't let us use the least noisy and be quiet about it? Differenting one *sin* from anther *sin* is just wasting time in hell.

Reply:
https://github.com/pmodels/mpich/pull/6390/commits/29fa9984146ef2444cffb30ec206a3d52164f8ee

Reply:
Looks like gcc won't complain casting from any function pointer to another, unlike the cast between different different pointer types. I guess they drew line where they think it gets too messy 😃